experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeBinder.ts

```diff
@@ -0,0 +1,930 @@
+				callback({
+					type: BindingType.Invalidation,
+				});
+			}
+		}
+		this.callbacks = undefined;
+		return this;
+	}
+}
+
+type CallableBindingContext = BindingContext & {
+	callbacks: Set<(...args: unknown[]) => any>;
+};
+
+class BufferingPathVisitor extends AbstractPathVisitor implements Flushable<BufferingPathVisitor> {
+	private readonly eventQueue: CallableBindingContext[] = [];
+
+	public constructor(options: BinderOptions) {
+		super(options);
+	}
+	public onDelete(path: UpPath, count: number): void {
+		const current = toDownPath<BindPath>(path);
+		const visitPaths = this.getRegisteredPaths(BindingType.Delete);
+		if (visitPaths !== undefined) {
+			for (const { path: visitPath, callbacks } of visitPaths) {
+				if (this.matchesPath(visitPath, current)) {
+					this.eventQueue.push({
+						path,
+						count,
+						type: BindingType.Delete,
+						callbacks,
+					});
+				}
+			}
+		}
+	}
+
+	public onInsert(path: UpPath, content: ProtoNodes): void {
+		const current = toDownPath<BindPath>(path);
+		const visitPaths = this.getRegisteredPaths(BindingType.Insert);
+		if (visitPaths !== undefined) {
+			for (const { path: visitPath, callbacks } of visitPaths) {
+				if (this.matchesPath(visitPath, current)) {
+					this.eventQueue.push({
+						path,
+						content,
+						type: BindingType.Insert,
+						callbacks,
+					});
+				}
+			}
+		}
+	}
+
+	public onSetValue(path: UpPath, value: TreeValue): void {
+		const current = toDownPath<BindPath>(path);
+		const visitPaths = this.getRegisteredPaths(BindingType.SetValue);
+		if (visitPaths !== undefined) {
+			for (const { path: visitPath, callbacks } of visitPaths) {
+				if (this.matchesPath(visitPath, current)) {
+					this.eventQueue.push({
+						path,
+						value,
+						type: BindingType.SetValue,
+						callbacks,
+					});
+				}
+			}
+		}
+	}
+
+	public flush(): BufferingPathVisitor {
+		const sortedQueue: CallableBindingContext[] = nativeSort(
+			this.eventQueue,
+			this.options.sortFn ?? compareBinderEventsDeleteFirst,
+		);
+		if (this.hasRegisteredContextType(BindingType.Batch)) {
+			const batchPaths = this.getRegisteredPaths(BindingType.Batch);
+			assert(batchPaths !== undefined, "batch paths confirmed registered");
+			const collected = new Set<(...args: unknown[]) => any>();
+			const batchEvents = sortedQueue.filter((event) => {
+				const current = toDownPath<BindPath>(event.path);
+				let filtered = false;
+				for (const { path: visitPath, callbacks } of batchPaths) {
+					if (this.matchesPath(visitPath, current)) {
+						callbacks.forEach((callback) => collected.add(callback));
+						filtered = true;
+						break;
+					}
+				}
+				return filtered;
+			});
+			if (batchEvents.length > 0) {
+				for (const callback of collected) {
+					callback({
+						type: BindingType.Batch,
+						events: batchEvents,
+					});
+				}
+			}
+			for (const event of batchEvents) {
+				const index = sortedQueue.indexOf(event);
```

Comment:
this is quite expensive, this algorithm will need n^2 operations if there are n batchEvent.

Reply:
Refactored, reduced the time complexity of processing batch events from O(n^2) to O(n).